### DISCUSSION TITLE
jython 스크립트로 테스트 중일 때, httprequest patch 사용하는 법?

### DISCUSSION THREAD

**Discussion by hoyongsh** [naver/ngrinder#820] [Q&A]
안녕하세요.

성능 테스트 진행 중 patch 호출 방법에 대해서 알고 싶어서 문의글 남깁니다.

```
from net.grinder.script import Test
from net.grinder.plugin.http import HTTPRequest
from net.grinder.plugin.http import HTTPPluginControl

request = HTTPRequest(url="domain")
result = request.POST(uri, strBody, header)
```

기존에는 이런 방식으로 호출을 했었는데, patch는 지원하지 않는다는것을 확인했습니다.
그리고 3.5.4 버전 이후부터, 새로운 http 클라이언트로 호출하여 patch가 지원한다는걸 확인했고, 
form-data 헤더 추가하여, 

```
import org.ngrinder.http.HTTPRequest
import org.ngrinder.http.HTTPResponse

org.ngrinder.http.HTTPRequest.PATCH("url",nvpair param,nvpair header)
```

 PATCH(): self arg can't be coerced to org.ngrinder.http.HTTPRequest, org.ngrinder.http.method.HTTPPatch [calling TestRunner]
net.grinder.scriptengine.jython.JythonScriptExecutionException: : PATCH(): self arg can't be coerced to org.ngrinder.http.HTTPRequest, org.ngrinder.http.method.HTTPPatch

이렇게 호출하고 있는데, 에러가 발생을 하고 있습니다.

방법이 잘못된것 같은데, 혹시 해결방법이 있을까요? 아니면 기존에 httprequest 에서 patch 메소드를 사용할 수 있는 방법이 있는지 궁금합니다.

**Comment by donggyu04** [comment]
왠만하면 groovy로 스크립트를 작성하시는걸 추천드립니다.
3.5.4 이상 버전에서는 groovy 스크립트 생성 시 기본적으로 신규 httpClient를 사용하도록 스크립트가 자동으로 구성됩니다.
[신규 HttpClient 가이드](https://github.com/naver/ngrinder/wiki/The-New-nGrinder-HTTP-Client)도 그루비로 작성 돼있으므로 참조하시기도 편할것 같구요.

Jython에서 신규 클라이언트를 사용하시려면 [신규 HttpClient 가이드](https://github.com/naver/ngrinder/wiki/The-New-nGrinder-HTTP-Client)를 참고 하셔서 관련 모듈을 메뉴얼하게 전부 import 하시고, 조금씩 변화가 있는 API요소들을 직접 컨버팅 해주셔야 되기 때문에 번거로우실 것 같습니다.

Jython에서 신규 클라이언트 사용시 파라미터 관련 오류가 있으시면 (i.e. Status 400 ) [HttpRequest](https://github.com/naver/ngrinder/blob/develop/ngrinder-runtime/src/main/java/org/ngrinder/http/HTTPRequest.java) 코드 참조 하셔서 스팩에 맞게 호출해 주시면 될 것 같습니다. 아래는 간단히 테스트 해본 샘플입니다. (쿠키 & HTTPRequestControl 설정 없음)

> 구 버전 클라이언트는 .PATCH() 형태로 호출 가능한 API를 제공하지 않는 것 같네요.

```python

from java.util import HashMap
from org.ngrinder.http import HTTPRequest, HTTPRequestControl, HTTPResponse

...

params = HashMap()
params["name"] = "Conor"

...

result = request.PATCH("{INPUT YOUR API URL}", params)

if result.statusCode == 200 :
	return
elif result.statusCode in (301, 302) :
	grinder.logger.warn("Warning. The response may not be correct. The response code was %d." %  result.statusCode)
	return
else :
	raise

```

**Reply by hoyongsh** [reply]
@donggyu04 

답변 감사합니다.
다음프로젝트부터는 groovy로 변경을 해야 될것 같네요.

혹시 jython스크립트로 이미지 업로드 테스트 할수 있는 방법은 있을까요?
resources 폴더안에 이미지 넣어 두고, 보낼때 PUT으로 보내려고 하는데 잘안되네요.

기존에는 
```
multipart_header = [NVPair('Content-Type', 'multipart/form-data;boundary=--------adfadsf----')]
imgfiles = (NVPair("idcard_img_path","./resources/appIcon.png"),)
param = ( NVPair("run number", str(grinder.runNumber)), )
strBody = (NVPair("id",nickName)
data = Codecs.mpFormDataEncode(strBody, imgfiles, multipart_header)
result = request.POST("uri",data,multipart_header)
```

이런 방식으로 보냈었는데, img.png를 byte[]로 가지고 와서 바로 
PUT(String uri, byte[] content, Map<String, String> headers)
이런방식으로 이미지를 보낼수 있는 방법이 있는지 궁금합니다.
